base-math/truncdddf.c

```diff
@@ -66,15 +66,25 @@ CONVERT_WRAPPER(
 	          return SIGNBIT(a) ? -INFINITY : INFINITY;
 	      }
 	  }
-	else if (exp < -BINPOWOF10_LIMIT)	/* Obvious underflow.  */
+	else if (exp <= -BINPOWOF10_DENORM_DBL)	/* Obvious underflow.  */
 	  {
 	    if (DFP_EXCEPTIONS_ENABLED)
 	      DFP_HANDLE_EXCEPTIONS (FE_UNDERFLOW|FE_INEXACT);
 
 	    switch (fegetround())
 	      {
 	        case FE_TONEAREST:
-	          return SIGNBIT(a) ? -__DBL_DENORM_MIN__ : __DBL_DENORM_MIN__;
+	          mant = llabs (a_norm);
+	          if (exp < -BINPOWOF10_DENORM_DBL ||
+	                    ((exp == -BINPOWOF10_DENORM_DBL) &&
+	                     (mant >= 24703282292062300)))
```

Comment:
Instead of comparing the exponent and mantissa, can't you just compare the absolute value against half of the minimum representable subnormal number, i.e.:
```
if (fabs (a) <= __DBL_DENORM_MIN__ / 2)
  return SIGNBIT (a) ? -0.0 : 0.0;
else
  return SIGNBIT (a) ? -__DBL_DENORM_MIN__ : __DBL_DENORM_MIN__;
```

You will not get rid of the need to define `BINPOWOF10_DENORM_DBL`, because of the previous use (in line 69), but you do get rid of this constant.

If it's not possible to do what I suggested, you should at least replace `24703282292062300` with a macro.

Reply:
We can't simple divide by two here because a^m/a^-n =  a^m-n the "half" here refer to the half of the mantissa. If the exponent is equal minimum subnormal (-324) and the mantissa higher than half of minimum subnormal mantissa, the nearest number will be 0.0 not the minimum subnormal all other case the nearest will be the minimum subnormal. If we divide by two it will divide the exponent too. I will think about a macro here.